Add or Update validation rule into the checkout address shipping and billing form fields (#9009)
magento/devdocs: src/guides/v2.3/howdoi/checkout/checkout_custom_validation.md

---
layout: tutorial
group: how-do-i
subgroup:
title: Add custom validation rules to checkout address form
subtitle: Customize Checkout
menu_order: 101
level3_subgroup: checkout-tutorial
functional_areas:
  - Checkout
---

> This topic describes how to add custom address validation in to the Magento checkout.  By default it has own set of validations for the address fields in the checkout step. But sometimes you may want to add some custom validations in to the different address fields. Some examples may be like limiting the character limit in the firstname and lastname of shipping and billing address.

Reviewer: ```suggestion
This topic describes how to add custom address validation to the Magento checkout.  By default it has own set of validations for the address fields in the checkout step. But sometimes you may want to add some custom validations to the different address fields. One example may be limiting the character limit in the firstname and lastname of the shipping and billing addresses.
```

> To change existing address validations to the Magento checkout, do the following steps:

Reviewer: ```suggestion
To change the existing address validations to the Magento checkout, do the following steps:
```

1. [Create a new module](#create-module).
1. [Create a di.xml file](#add-dixml-file).

> 1. [Adding validation to Shipping Address](#add-shipping-validation).

Reviewer: ```suggestion
1. [Adding validation to the Shipping Address](#add-shipping-validation).
```

> 1. [Adding validation to Billing Address](#add-billing-validation).

Reviewer: ```suggestion
1. [Adding validation to the Billing Address](#add-billing-validation).
```

## Step 1: Create a new module {#create-module}

[Create a new module](https://devdocs.magento.com/videos/fundamentals/create-a-new-module/) named `Vendor/AddressRestriction` and register it.

## Step 2 Create a di.xml file {#add-dixml-file}

Create the file `app/code/Vendor/AddressRestriction/etc/di.xml`:

```php?start_inline=1
<?xml version="1.0"?>
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:ObjectManager/etc/config.xsd">
    <type name="Magento\Checkout\Block\Checkout\LayoutProcessor">
        <plugin name="Vendor_AddressRestriction_Checkout" type="Vendor\AddressRestriction\Block\LayoutProcessor" sortOrder="100"/>
    </type>
</config>
```

> ## Step 3 Adding validation to Shipping Address {#add-shipping-validation}

Reviewer: ```suggestion
## Step 3 Adding validation to the Shipping Address {#add-shipping-validation}
```

Author: @BarnyShergold Thanks for your review and suggestion. All the suggestions are done please review it.

## Step 2 Create a di.xml file {#add-dixml-file}

Create the file `app/code/Vendor/AddressRestriction/etc/di.xml`:

```php?start_inline=1
<?xml version="1.0"?>
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:ObjectManager/etc/config.xsd">
    <type name="Magento\Checkout\Block\Checkout\LayoutProcessor">
        <plugin name="Vendor_AddressRestriction_Checkout" type="Vendor\AddressRestriction\Block\LayoutProcessor" sortOrder="100"/>
    </type>
</config>
```

## Step 3 Adding validation to the Shipping Address {#add-shipping-validation}

Create the file `app/code/Vendor/AddressRestriction/Block/LayoutProcessor.php`;

```php?start_inline=1
<?php
namespace Vendor\AddressRestriction\Block;

class LayoutProcessor
{
    /**
    * @param \Magento\Checkout\Block\Checkout\LayoutProcessor $subject
    * @param array $jsLayout
    * @return array
    */
    public function afterProcess(
    \Magento\Checkout\Block\Checkout\LayoutProcessor $subject,
    array $jsLayout
    ) {
        // Firstname

>         $jsLayout['components']['checkout']['children']['steps']['children']['shipping-step']['children']

Reviewer 2: In my opinion it's much better to suggest using the [Array Manager](https://devdocs.magento.com/guides/v2.4/extension-dev-guide/framework/array-manager.html), which is much preferred way of accessing a chained array. As result we'll have a much cleaner solution and code.

Author: @eduard13 Thanks for your review and suggestion. I have updated the mentioned suggestion. Kindly look into that.

Reviewer 2: @rajesh-mage, looks like you didn't get me right. I suggest using the Array Manager to access the array properties, please check the 1st example in the doc's article.

Please let me know if there are any concerns or questions.

        // Country
        $fields = $jsLayout['components']['checkout']['children']['steps']['children']['shipping-step']['children']
                ['shippingAddress']['children']['shipping-address-fieldset']['children']['country_id']['validation']['max_text_length'] = 31;

        // Postcode
        $fields = $jsLayout['components']['checkout']['children']['steps']['children']['shipping-step']['children']
                ['shippingAddress']['children']['shipping-address-fieldset']['children']['postcode']['validation']['max_text_length'] = 13;

        // Street address
        $fields = $jsLayout['components']['checkout']['children']['steps']['children']['shipping-step']['children']
                ['shippingAddress']['children']['shipping-address-fieldset']
                ['children']['street']['children'][0]['validation']['max_text_length'] = 40;

        return $fields;
    }
}
```

## Step 4: Add validation for the billing address {#add-billing-validation}

```php?start_inline=1
$fields = $jsLayout['components']['checkout']['children']['steps']['children']['billing-step']['children']

>         ['shippingAddress']['children']['shipping-address-fieldset']['children']['firstname']['validation']['max_text_length'] = 20;

Reviewer 3: Are all these references to the shipping address applicable for validating the billing address? Or is this doing something to ensure the values are the same for both types of addresses?
